File: server/backup_database.py
#!/usr/bin/env python3

import argparse
import sqlite3
from datetime import datetime, timezone
from pathlib import Path


def backup_database(database: Path, output_dir: Path, keep: int = 14) -> Path:
    if keep < 1:
        raise ValueError("keep must be at least 1")
    output_dir.mkdir(parents=True, exist_ok=True)
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S%fZ")
    destination = output_dir / f"gateway-{timestamp}.db"
    temporary = destination.with_suffix(".db.partial")
    try:
        with sqlite3.connect(database) as source, sqlite3.connect(
            temporary
        ) as target:
            source.backup(target)
            if target.execute("PRAGMA integrity_check").fetchone()[0] != "ok":
                raise RuntimeError("backup integrity check failed")
        temporary.chmod(0o600)
        temporary.replace(destination)
    finally:
        if temporary.exists():
            temporary.unlink()
    backups = sorted(
        output_dir.glob("gateway-*.db"),
        key=lambda path: path.stat().st_mtime,
        reverse=True,
    )
    for stale in backups[keep:]:
        stale.unlink()
    return destination


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Create a consistent SQLite gateway backup."
    )
    parser.add_argument(
        "--database",
        type=Path,
        default=Path("/var/lib/gateway/gateway.db"),
    )
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=Path("/var/lib/gateway/backups"),
    )
    parser.add_argument("--keep", type=int, default=14)
    args = parser.parse_args()
    backup = backup_database(args.database, args.output_dir, args.keep)
    print(f"Backup created: {backup}")


if __name__ == "__main__":
    main()

File: server-node/src/operations/backup-database.ts
import { backup, DatabaseSync } from 'node:sqlite';
import { chmodSync, mkdirSync, readdirSync, renameSync, rmSync, statSync } from 'node:fs';
import { basename, join } from 'node:path';

export async function backupDatabase(databasePath: string, outputDirectory: string, keep: number): Promise<string> {
  if (!Number.isInteger(keep) || keep < 1) {
    throw new Error('keep must be at least 1');
  }
  mkdirSync(outputDirectory, { recursive: true, mode: 0o700 });
  const timestamp = new Date().toISOString().replace(/[-:.]/g, '').replace('Z', '000Z');
  const destination = join(outputDirectory, `gateway-${timestamp}.db`);
  const temporary = `${destination}.partial`;
  rmSync(temporary, { force: true });

  const source = new DatabaseSync(databasePath, { readOnly: true });
  try {
    await backup(source, temporary);
  } finally {
    source.close();
  }

  try {
    finalizeStandaloneDatabase(temporary);
    chmodSync(temporary, 0o600);
    renameSync(temporary, destination);
  } finally {
    removeDatabaseFiles(temporary);
  }

  const backups = readdirSync(outputDirectory)
    .filter(name => /^gateway-.*\.db$/.test(name))
    .map(name => join(outputDirectory, name))
    .sort((left, right) => statSync(right).mtimeMs - statSync(left).mtimeMs);
  for (const stale of backups.slice(keep)) rmSync(stale);
  for (const name of readdirSync(outputDirectory)) {
    if (/^gateway-.*\.db(?:\.partial)?-(?:wal|shm)$/.test(name)) {
      rmSync(join(outputDirectory, name), { force: true });
    }
  }
  return destination;
}

function finalizeStandaloneDatabase(path: string): void {
  const database = new DatabaseSync(path);
  try {
    database.exec('PRAGMA wal_checkpoint(TRUNCATE)');
    database.exec('PRAGMA journal_mode = DELETE');
    const row = database.prepare('PRAGMA integrity_check').get() as { integrity_check?: string };
    if (row.integrity_check !== 'ok') throw new Error('backup integrity check failed');
  } finally {
    database.close();
  }
  rmSync(`${path}-wal`, { force: true });
  rmSync(`${path}-shm`, { force: true });
}

function removeDatabaseFiles(path: string): void {
  rmSync(path, { force: true });
  rmSync(`${path}-wal`, { force: true });
  rmSync(`${path}-shm`, { force: true });
}

async function main(): Promise<void> {
  const args = parseArguments(process.argv.slice(2));
  const path = await backupDatabase(args.database, args.outputDirectory, args.keep);
  console.log(`Backup created: ${path}`);
}

function parseArguments(argv: string[]): { database: string; outputDirectory: string; keep: number } {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    const key = argv[index];
    const value = argv[index + 1];
    if (!key?.startsWith('--') || value === undefined) throw new Error('invalid arguments');
    values.set(key, value);
  }
  const keepRaw = values.get('--keep') ?? '14';
  if (!/^\d+$/.test(keepRaw)) throw new Error('keep must be at least 1');
  return {
    database: values.get('--database') ?? '/var/lib/gateway/gateway.db',
    outputDirectory: values.get('--output-dir') ?? '/var/lib/gateway/backups',
    keep: Number(keepRaw),
  };
}

if (basename(process.argv[1] ?? '') === 'backup-database.js') {
  main().catch(error => {
    console.error(error instanceof Error ? error.message : error);
    process.exitCode = 1;
  });
}
